packages/flame/lib/src/experimental/geometry/shapes/polygon.dart

```diff
@@ -0,0 +1,219 @@
+import 'dart:ui';
+
+import 'package:vector_math/vector_math_64.dart';
+
+import '../../../game/transform2d.dart';
+import 'shape.dart';
+
+/// An arbitrary polygon with 3 or more vertices.
+///
+/// The vertices of the polygon are in the counter-clockwise order, so that the
+/// polygon's interior is in your left-hand side as you traverse the polygon's
+/// vertices.
+///
+/// A polygon can be either convex or not, the [containsPoint] method will work
+/// in both cases, however, the method used with convex polygon is faster.
+class Polygon extends Shape {
+  /// Constructs the polygon from the given list of vertices.
+  ///
+  /// If the list is not in the counter-clockwise order, then it will be
+  /// reversed in-place.
+  ///
+  /// If the [convex] flag is provided, then it serves as a hint about whether
+  /// the polygon is convex or not. With this flag the user promises that the
+  /// vertices are already in the correct CCW order.
+  Polygon(this._vertices, {bool? convex})
+      : assert(_vertices.length >= 3, 'At least 3 vertices are required') {
+    _initializeEdges();
+    if (convex == null) {
+      _checkOrientation();
+    } else {
+      _convex = convex;
+    }
+  }
+
+  /// The vertices (corners) of the polygon.
+  ///
+  /// The user should treat this list as read-only and not attempt to modify
+  /// either the list itself or individual points.
+  List<Vector2> get vertices => _vertices;
```

Comment:
This could return an `UnmodifiableList` maybe?

Reply:
I don't think I like `UnmodifiableList`, it looks rather inefficient.